src/coreclr/src/System.Private.CoreLib/src/System/StubHelpers.cs

```diff
@@ -686,11 +686,61 @@ internal static long ConvertToManaged(double nativeDate)
 #if FEATURE_COMINTEROP
     internal static class InterfaceMarshaler
     {
+        // See interopconverter.h
+        [Flags]
+        private enum ItfMarshalFlags
+        {
+            ITF_MARSHAL_INSP_ITF        = 0x01,
+            ITF_MARSHAL_SUPPRESS_ADDREF = 0x02,
+            ITF_MARSHAL_CLASS_IS_HINT   = 0x04,
+            ITF_MARSHAL_DISP_ITF        = 0x08,
+            ITF_MARSHAL_USE_BASIC_ITF   = 0x10,
+            ITF_MARSHAL_WINRT_SCENARIO  = 0x20,
+        };
+
+        private static bool IsForIUnknown(int flags)
+        {
+            ItfMarshalFlags interfaceFlags = (ItfMarshalFlags)flags;
+            return (interfaceFlags & ItfMarshalFlags.ITF_MARSHAL_USE_BASIC_ITF) != 0
+                && (interfaceFlags & ItfMarshalFlags.ITF_MARSHAL_INSP_ITF) == 0
+                && (interfaceFlags & ItfMarshalFlags.ITF_MARSHAL_DISP_ITF) == 0;
+        }
+
+        internal static IntPtr ConvertToNative(object objSrc, IntPtr itfMT, IntPtr classMT, int flags)
+        {
+            if (ComWrappers.IsGlobalInstanceRegistered() && IsForIUnknown(flags))
```

Comment:
Is it possible to handle all cases? I think this needs to handle all cases in order to be useful.

For example, I can imagine this being useful for statically pre-compiled COM interop carried with the apps.

Reply:
>  I can imagine this being useful for statically pre-compiled COM interop carried with the apps.

This may be a good test case: Take a test or an app that does a bunch of COM interop, and inject the COM interop into it via this so that the build-in COM interop is not used.

Reply:
If this all encompassing approach is being done now - was hoping for post .NET 5 - then we should also update [`Marshal.GetIDispatchForObject()`](https://docs.microsoft.com/dotnet/api/system.runtime.interopservices.marshal.getidispatchforobject).

Reply:
Doing this after .NET 5 would be a breaking change, so we would need to think twice about the breaking potential, etc.

It would be best to try to make this work the way we want for .NET 5.

Reply:
I was thinking of the next major version (e.g. .NET 6). I appreciate the breaking change concern, which is why I thought we were limiting this to just the CsWinRT scenarios since learnings would fall out of that and we could fold that into general support. My concern with this now is merely the lack of concrete best practices and we will probably need enhancements (e.g. new enums values) - without some real world learnings I'm not confident the API is ready.

That could be me just being shy of providing such a base building block, but I don't really see the need to get it completely wired up when the only scenario that really needs it is CsWinRT for .NET 5.

Not say that we can't wire all it up, but I don't see the pressing concern to push something right now.

Reply:
It is going to be a public API. It is likely that other people are going to show up who will start using it.

If we are not confident in the shape and the behavior of the API, we should ship it as experimental (e.g. similar to Utf8 string) so that it is not present in the officially supported public surface.

Reply:
Let's just do it. I am not sure how to convey my concerns properly.

Reply:
Agree with all options except for the `IDispatch` support. We shouldn't expect the `ComWrappers` to support that yet since it is so onerous. Perhaps we can address that in the future but for right now I would prefer if we at least block that support.

Reply:
Updated to basically do this for everything now (including COM activation) if there is a `ComWrappers` registered globally.